// @pepperi-addons/papi-sdk/dist/endpoints/codeJobs.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodeJobsEndpoint = void 0;
const endpoint_1 = __importDefault(require("../endpoint"));
class CodeJobEndpoint {
    constructor(service, uuid, async) {
        this.service = service;
        this.uuid = uuid;
        this.async = async;
    }
    async find(includeDeleted = false) {
        return await this.get(includeDeleted);
    }
    async get(includeDeleted = false) {
        return await this.service.get(`/code_jobs/${this.uuid}?include_deleted=${includeDeleted}`);
    }
    async publish(body) {
        return this.service.post(`/code_jobs/${this.uuid}/publish`, body);
    }
    async execute() {
        const asyncPart = this.async ? 'async/' : '';
        return this.service.post(`/code_jobs/${asyncPart}${this.uuid}/execute`);
    }
}
class CodeJobsEndpoint extends endpoint_1.default {
    constructor(service) {
        super(service, '/code_jobs');
        this.isAsync = false;
    }
    uuid(uuid) {
        return new CodeJobEndpoint(this.service, uuid, this.isAsync);
    }
    async() {
        this.isAsync = true;
        return this;
    }
}
exports.CodeJobsEndpoint = CodeJobsEndpoint;
