// @pepperi-addons/papi-sdk/dist/entities/data-view.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataViewRowModes = exports.HorizontalAlignments = exports.VerticalAlignments = exports.ResoursePrefixes = exports.ResourceTypes = exports.DataViewFieldTypes = exports.DataViewScreenSizes = exports.DataViewTypes = void 0;
exports.DataViewTypes = [
    'Grid',
    'Form',
    'Card',
    'Large',
    'Line',
    'Map',
    'Menu',
    'Configuration',
    'CardsGrid',
    'Details',
];
exports.DataViewScreenSizes = ['Tablet', 'Phablet', 'Landscape'];
exports.DataViewFieldTypes = {
    None: 0,
    TextBox: 1,
    LimitedLengthTextBox: 2,
    TextArea: 3,
    TextHeader: 4,
    Date: 5,
    DateAndTime: 6,
    NumberInteger: 7,
    NumberReal: 8,
    Currency: 9,
    Boolean: 10,
    ComboBox: 11,
    MultiTickBox: 12,
    Separator: 13,
    Address: 14,
    Percentage: 15,
    EmptyComboBox: 16,
    InternalLink: 17,
    Email: 18,
    LimitedDate: 19,
    Image: 20,
    MultiTickBoxToComboBox: 21,
    EmptyMultiTickBox: 22,
    Totals: 23,
    Attachment: 24,
    Signature: 25,
    Link: 26,
    ImageURL: 27,
    NumberIntegerQuantitySelector: 28,
    NumberRealQuantitySelector: 29,
    NumberIntegerForMatrix: 30,
    NumberRealForMatrix: 31,
    Images: 32,
    Indicators: 33,
    CalculatedReal: 34,
    CalculatedInt: 35,
    CalculatedString: 36,
    CalculatedDate: 37,
    CalculatedBool: 38,
    MapDataDropDown: 39,
    MapDataReal: 40,
    MapDataString: 41,
    MapDataInt: 42,
    Sum: 43,
    Phone: 44,
    UrlForApi: 45,
    ManyToManyUrlForApi: 46,
    ReferenceType: 47,
    GuidReferenceType: 48,
    Button: 49,
    UIControlFieldType_InternalPage: 50,
    Duration: 51,
    ListOfObjects: 52,
    Package: 53,
    RelatedObjectsCards: 54,
    BooleanText: 55,
    RichTextHTML: 56,
};
exports.ResourceTypes = [
    'None',
    'accounts',
    'transactions',
    'activities',
    'all_activities',
    'items',
    'users',
    'transaction_lines',
    'contacts',
    'lists',
    'catalogs',
    'types',
];
exports.ResoursePrefixes = ['GA', 'OA', 'CP', 'AT', 'GL', 'CA'];
exports.VerticalAlignments = {
    Stretch: 0,
    Top: 1,
    Bottom: 2,
    Center: 3,
};
exports.HorizontalAlignments = {
    Stretch: 0,
    Left: 1,
    Right: 2,
    Center: 3,
};
exports.DataViewRowModes = {
    Fixed: 0,
    MatchParent: 1,
};
